onnxruntime/core/providers/rocm/fpgeneric.cu

```diff
@@ -0,0 +1,162 @@
+//   int id = blockIdx.x * blockDim.x + threadIdx.x;
+//   if (id >= n) return;
+
+//   hiprandState localState = *state;
+
+//   float x;
+//   skipahead(id, &localState);
+//   x = hiprand_uniform(&localState);
+
+//   result[id] = x;
+//   if (id == n - 1) *state = localState;
+// }
+
+// __global__ void GenerateNormalHalf(hiprandState* state, half* result, int n, half mean, half stddev) {
+//   int id = blockIdx.x * blockDim.x + threadIdx.x;
+//   if (id >= n) return;
+
+//   hiprandState localState = *state;
+
+//   float x;
+//   skipahead(id, &localState);
+//   x = hiprand_normal(&localState);
+
+//   result[id] = (float)mean + (float)stddev * x;
+//   if (id == n - 1) *state = localState;
+// }
+
+// kernels can convert matrix between half and float. speed currently not optimized, may need to add half2
+/*
+__global__ void copyHalf2Float(float *odata, const half *idata, const int n)
+{
+    float tmp[COPY_TILE_DIM/COPY_BLOCK_DIM];
+
+    int x = blockIdx.x * COPY_TILE_DIM + threadIdx.x;
+
+    for (int j = 0; j < COPY_TILE_DIM/COPY_BLOCK_DIM; j++)
+        tmp[j] = (float) idata[x + j*COPY_BLOCK_DIM];
+
+    for (int j = 0; j < COPY_TILE_DIM/COPY_BLOCK_DIM; j++)
+        if(x + j*COPY_BLOCK_DIM < n) odata[x + j*COPY_BLOCK_DIM] = tmp[j];
+}
+
+__global__ void copyFloat2Half(half *odata, const float *idata, const int n)
+{
+    float tmp[COPY_TILE_DIM/COPY_BLOCK_DIM];
+
+    int x = blockIdx.x * COPY_TILE_DIM + threadIdx.x;
+
+    for (int j = 0; j < COPY_TILE_DIM/COPY_BLOCK_DIM; j++)
+        tmp[j] = idata[x + j*COPY_BLOCK_DIM];
+
+    for (int j = 0; j < COPY_TILE_DIM/COPY_BLOCK_DIM; j++)
+        if(x + j*COPY_BLOCK_DIM < n) odata[x + j*COPY_BLOCK_DIM] = tmp[j];
+}
+*/
+
+__global__ void CopyVectorHalf(const half* x, int incx, half* y, int incy, int n) {
+  int id = blockIdx.x * blockDim.x + threadIdx.x;
+  if (id >= n) return;
+  y[id * incy] = x[id * incx];
+}
+
+}  // namespace
+
+hipblasStatus_t hipblasTransposeHelper(hipblasHandle_t, hipblasOperation_t, hipblasOperation_t, int m, int n, const half*, const half* A, int, const half*, const half*, int, half* C, int) {
+  if (C != A) {
+    dim3 dimGrid((n + TRANS_TILE_DIM - 1) / TRANS_TILE_DIM, (m + TRANS_TILE_DIM - 1) / TRANS_TILE_DIM, 1);
+    dim3 dimBlock(TRANS_TILE_DIM, BLOCK_ROWS, 1);
+
+    hipLaunchKernelGGL(transposeNoOverlap, dim3(dimGrid), dim3(dimBlock), 0, 0, C, A, n, m);
+  } else {
+    return HIPBLAS_STATUS_NOT_SUPPORTED;
+  }
+  return HIPBLAS_STATUS_SUCCESS;
+}
+
+hipblasStatus_t hipblasCopyHelper(hipblasHandle_t, int n, const half* x, int incx, half* y, int incy) {
+  dim3 dimGrid((unsigned int)(n + COPY_BLOCK_DIM - 1) / COPY_BLOCK_DIM, 1, 1);
+  dim3 dimBlock(COPY_BLOCK_DIM, 1, 1);
+  hipLaunchKernelGGL(CopyVectorHalf, dim3(dimGrid), dim3(dimBlock), 0, 0, x, incx, y, incy, n);
+  return HIPBLAS_STATUS_SUCCESS;
+}
+
+// hiprandStatus_t curandGenerateUniformHelper(hiprandGenerator_t, half* outputPtr, size_t num) {
```

Comment:
This is uncommented for CUDA case?

Reply:
These code needs to be cleaned later since they are not used in CUDA EP as well.